Implement index adjustment for rel json ptr

handrews opened the pull request:
Fixes #23 : Adds support for index adjustment to class `RelativeJSONPointer` per the [most recent draft](https://datatracker.ietf.org/doc/html/draft-bhutton-relative-json-pointer-00), as modified by [recently proposed fixes](https://github.com/handrews/json-schema-spec/compare/main...handrews:json-schema-spec:relptr-more-fixes) that correct errors in the original wording for index adjustment.

@marksparkza I've tried to match styles, but I'm perfectly happy to change formatting or wording or whatever so feel free to nitpick.  I've not used type hinting in Python before so if I missed something obvious that I should have done with types it's because I don't know what I'm doing :-)

For the tests, I matched the sort of tests that were already present.  I noticed some things that were not covered, such as some syntax error conditions, but I have filed those ideas as issue #48.

I started writing a tutorial page for `RelativeJSONPointer`, but decided to submit it as a separate PR.

With the adjustment to hypothesis 6.0.3 per #45 , all tests run and pass on Python 3.8, 3.9, 3.10, and 3.11.

**jsonpointer.py coverage HTML page header _before_:**
```
Coverage for jschon/jsonpointer.py: 93%
143 statements    134 run    9 missing    11 excluded    4 partial
```

**jsonpointer.py coverage HTML page header _after_:**
```
Coverage for jschon/jsonpointer.py: 96%
143 statements    137 run    6 missing    11 excluded    3 partial
```

**Test & coverage report _before_**:
```
------------------------------------------------------------------------------------------------- benchmark: 5 tests -------------------------------------------------------------------------------------------------
Name (time in us)                      Min                    Max                  Mean                StdDev                Median                 IQR            Outliers          OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_create_json                   40.0420 (1.0)      22,765.2910 (49.10)       55.9223 (1.0)        241.7498 (6.51)        41.2089 (1.0)        2.4159 (1.0)        2;2022  17,881.9634 (1.0)        8983           1
test_evaluate_json[valid]         151.0840 (3.77)     15,631.1250 (33.72)      172.3141 (3.08)       256.0344 (6.89)       157.4580 (3.82)       4.5430 (1.88)        1;583   5,803.3546 (0.32)       3702           1
test_evaluate_json[invalid]       180.6669 (4.51)        463.6250 (1.0)        199.4462 (3.57)        37.1391 (1.0)        186.0000 (4.51)       4.7920 (1.98)      558;601   5,013.8835 (0.28)       3268           1
test_create_schema                481.7080 (12.03)    25,930.1249 (55.93)      546.1360 (9.77)       898.4470 (24.19)      496.2500 (12.04)     22.5734 (9.34)         3;43   1,831.0458 (0.10)       1463           1
test_validate_schema            2,487.3330 (62.12)    43,008.5000 (92.77)    3,375.4863 (60.36)    4,836.1567 (130.22)   2,542.6251 (61.70)    635.9163 (263.22)        6;6     296.2536 (0.02)        367           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
========================================================================= 2396 passed in 113.10s (0:01:53) ==========================================================================
py310: commands_post[0]> coverage report
Name                              Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------------------
jschon/__init__.py                   20      2      2      0    91%
jschon/catalog/_2019_09.py           18      0      0      0   100%
jschon/catalog/_2020_12.py           18      0      0      0   100%
jschon/catalog/__init__.py          136     11     30      3    89%
jschon/catalog/_next.py              18      0      0      0   100%
jschon/exceptions.py                 14      0      0      0   100%
jschon/formats.py                     7      0      4      1    91%
jschon/json.py                      210     11     86      4    95%
jschon/jsonpatch.py                 169     43     60     15    70%
jschon/jsonpointer.py               128      6     40      2    95%
jschon/jsonschema.py                263     18    121     11    90%
jschon/output.py                     80      2     41      1    96%
jschon/uri.py                        68      2     14      2    95%
jschon/utils.py                      40      5     12      2    87%
jschon/vocabulary/__init__.py        69      6     24      5    86%
jschon/vocabulary/annotation.py      36      0      2      0   100%
jschon/vocabulary/applicator.py     254      0    147      2    99%
jschon/vocabulary/core.py           123     14     36      8    85%
jschon/vocabulary/format.py          28      0      4      0   100%
jschon/vocabulary/legacy.py         100      2     58      5    96%
jschon/vocabulary/validation.py     159      0     68      1    99%
-------------------------------------------------------------------
TOTAL                              1958    122    749     62    92%
```

**Test & coverage report _after_:**
```
------------------------------------------------------------------------------------------------- benchmark: 5 tests -------------------------------------------------------------------------------------------------
Name (time in us)                      Min                    Max                  Mean                StdDev                Median                 IQR            Outliers          OPS            Rounds  Iterations
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
test_create_json                   40.5411 (1.0)         735.0410 (1.52)        55.1769 (1.0)         36.5315 (1.10)        42.3340 (1.0)        4.1666 (1.0)        61;494  18,123.5242 (1.0)        2164           1
test_evaluate_json[valid]         153.1660 (3.78)        482.3339 (1.0)        170.9793 (3.10)        33.1258 (1.0)        159.5829 (3.77)       4.5832 (1.10)      631;686   5,848.6608 (0.32)       4201           1
test_evaluate_json[invalid]       181.4590 (4.48)     15,583.5000 (32.31)      203.9511 (3.70)       244.8774 (7.39)       187.1660 (4.42)       5.1241 (1.23)       18;743   4,903.1356 (0.27)       4034           1
test_create_schema                482.4999 (11.90)    26,491.4590 (54.92)      545.4101 (9.88)       852.8854 (25.75)      498.5830 (11.78)     23.8648 (5.73)         3;72   1,833.4826 (0.10)       1667           1
test_validate_schema            2,493.6250 (61.51)    46,667.7080 (96.75)    3,446.7382 (62.47)    5,215.5207 (157.45)   2,557.6040 (60.41)    638.9170 (153.34)        6;6     290.1294 (0.02)        360           1
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

Legend:
  Outliers: 1 Standard Deviation from Mean; 1.5 IQR (InterQuartile Range) from 1st Quartile and 3rd Quartile.
  OPS: Operations Per Second, computed as 1 / Mean
========================================================================= 2402 passed in 123.74s (0:02:03) ==========================================================================
py310: commands_post[0]> coverage report
Name                              Stmts   Miss Branch BrPart  Cover
-------------------------------------------------------------------
jschon/__init__.py                   20      2      2      0    91%
jschon/catalog/_2019_09.py           18      0      0      0   100%
jschon/catalog/_2020_12.py           18      0      0      0   100%
jschon/catalog/__init__.py          136     11     30      3    89%
jschon/catalog/_next.py              18      0      0      0   100%
jschon/exceptions.py                 14      0      0      0   100%
jschon/formats.py                     7      0      4      1    91%
jschon/json.py                      210     11     86      4    95%
jschon/jsonpatch.py                 169     43     60     15    70%
jschon/jsonpointer.py               143      6     50      3    95%
jschon/jsonschema.py                263     18    121     11    90%
jschon/output.py                     80      2     41      1    96%
jschon/uri.py                        68      2     14      2    95%
jschon/utils.py                      40      5     12      2    87%
jschon/vocabulary/__init__.py        69      6     24      5    86%
jschon/vocabulary/annotation.py      36      0      2      0   100%
jschon/vocabulary/applicator.py     254      0    147      2    99%
jschon/vocabulary/core.py           123     14     36      8    85%
jschon/vocabulary/format.py          28      0      4      0   100%
jschon/vocabulary/legacy.py         100      2     58      5    96%
jschon/vocabulary/validation.py     159      0     68      1    99%
-------------------------------------------------------------------
TOTAL                              1973    122    759     63    92%
```

handrews commented:
It looks like the tests being run on this change are hitting the issue that PR #49 works around.

marksparkza commented on jschon/jsonpointer.py:
```diff
@@ -272,6 +278,10 @@ def __new__(
+        :param over: the integer value used to adjust the array index after
+            applying `up`, which is only valid if that location is an array item;
+            a value of 0, which is not allowed by the grammar, is treated as if
+            there is no adjutsment.
```
typo

marksparkza commented on jschon/jsonpointer.py:
```diff
@@ -282,7 +292,14 @@ def __new__(
-            up, ref = match.group('up', 'ref')
+            up, over, ref = match.group('up', 'over', 'ref')
+
+        if over in (0, ''):
```
I'd suggest `if not over` which succinctly passes `None`, `0` or `''`.

Declaring the parameter as `over: int = None` might concur more strictly with the spec, but I think this would be overkill since, in the context of index manipulation, `None` implies `0`. In any case, `if not over` allows the caller to obtain the `over` value from a `None`, `0` or `''`, which are essentially equivalent.

handrews replied:
Yes, all of that makes sense (and I like how it makes the test code more straightforward as well).

I originally had it as `over: int = None` but the way I was handling it got very fiddly.

Would it make senes to use `over: int = ''` because that's how things will come in by default from parsing as seen in the test case?  It just felt weird to do that.  `over: int = None` seemed OK on the principle that something can be type-or-`None`, but type-or-other-type seemed too weird.  But I don't really understand type hinting and what constitutes a reasonable vs unreasonable violation of the hint.

handrews replied:
I did want to allow `0` as a passed value, because logically it is supported, it is just not _visible_ in the string representation of the pointer.  And that way if you're doing math to calculate the adjustment you don't have to worry about converting `0` to `None` (or `''`).

marksparkza replied:
In hindsight allowing strings to be passed as kwargs was a mistake. It complicates the code and presents a confusing interface to callers - the caller can already pass the whole rjp string via `value`, so why would they want to split that up and pass component strings via kwargs? In fact, the only reason this was done was to permit laziness in the unit test. I'm going to proceed with merging the PR, but will create a subsequent commit to resolve this.

handrews replied:
@marksparkza yeah that makes sense to me!

marksparkza commented on jschon/jsonpointer.py:
```diff
@@ -323,6 +341,16 @@ def evaluate(self, document: JSON) -> Union[int, str, JSON]:
+            if node.parent.type != "array":
+                raise RelativeJSONPointerError(f'Index adjustment not valid for type {node.parent.type}')
+            adjusted = int(node.key) + self.over
+            if adjusted < 0 or adjusted > len(node.parent):
```
`adjusted < 0 or adjusted >= len(node.parent)`

marksparkza commented on tests/data/relative_jsonpointer.json:
```diff
@@ -217,5 +217,44 @@
+            },
+            {
+                "start": "/2/baz",
+                "ref": "1+5",
```
I'd suggest changing this to catch the edge case: 
`"ref": "1+1"`

handrews replied:
Nice catch (here and in the code).  As a former QA automation engineer, I'm a touch embarrassed to have overlooked it! 😅 Just proves the rule that devs miss things in our own code.

marksparkza commented on tests/data/relative_jsonpointer.json:
```diff
@@ -217,5 +217,44 @@
+                "ref": "0+1",
+                "result": "<fail>"
+            }
+        ]
```
Maybe add one test for a negative adjustment error, like:
```
            {
                "start": "/1",
                "ref": "0-2",
                "result": "<fail>"
            }
```

marksparkza commented on tests/test_jsonpointer.py:
```diff
@@ -123,11 +123,19 @@ def test_compare_jsonpointer(value1, value2):
     match = re.fullmatch(relative_jsonpointer_regex, value)
-    up, ref = match.group('up', 'ref')
+    up, over, ref = match.group('up', 'over', 'ref')
+    kwargs = {
+        'up': up,
+        'ref': ref,
+    }
+    if over:
+        kwargs['over'] = over
```
```py
    kwargs = {
        'up': up,
        'over': over,
        'ref': ref,
    }
```
suffices here, since we're handling `''` as if it were `0`.

marksparkza commented on tests/test_jsonpointer.py:
```diff
@@ -123,11 +123,19 @@ def test_compare_jsonpointer(value1, value2):
+    r2 = RelativeJSONPointer(**kwargs)
     assert r1 == r2
     assert str(r1) == value
+    assert eval(repr(r1)) == r1
```
:tada:

marksparkza commented:
> @marksparkza I've tried to match styles

`up` and `over`. I like it. This kind of style works in jschon :smile:  I'm not a huge fan of rigid formatting a la black. Elegance and clarity are paramount.

handrews commented:
@marksparkza I've incorporated your feedback, rebased on top of the workaround for the testing issue, and also fixed a bug I noticed in the test case for creating relative JSON pointers — `0` vs `'0'` — and reorganized it a bit for more thorough testing of default values and `JSONPointer` instances.
